### PR TITLE
Preserve discarded effects and optimize stackified bytecode

### PR DESCRIPTION
**Pull request by antoniosarosi**
## Summary

Fix discarded-expression correctness first, then optimize MIR values and stackified bytecode.

This PR has three commits (the third is the requested follow-up for inlined binary prefixes):

1. `fix(compiler): preserve effects of discarded expressions`
2. `perf(compiler): optimize MIR values and stackified bytecode`
3. `perf(compiler): carry binary prefixes through inlined expressions`

The second commit:

- Gives short-circuit expressions their own MIR result, independent of mutable destination locals, and represents null coalescing explicitly.
- Lowers `&&`, `||`, and `!` in predicates directly to CFG branches, preserving non-boolean truthiness conversion.
- Adds O2 scalar constant propagation/folding and boolean-branch simplification, plus definition-level dead-store elimination with exception-handler liveness.
- Chooses branch polarity from emitted layout and threads identical value-preserving branches.
- Extends validated stack carry to virtual-call results, call arguments, binary operands, and mixed constant/call prefixes. Requires unique definitions and single-entry continuations, and revalidates the final carry plan.
- Preserves the existing phi-coverage checks and O0 named-scalar slot behavior. This is an incremental MIR improvement, not a wholesale SSA rewrite.

### Discarded short circuits dropped conditional RHS effects

```baml
class Box { value: int }

function and_inline(a: bool, box: Box) -> int {
    a && { box.value = box.value + 1; true }
    box.value
}

test "discarded result must not discard the mutation" {
    assert.equal(and_inline(true, Box { value: 0 }), 1)
    assert.equal(and_inline(false, Box { value: 0 }), 0)
}
```

Before this PR, the first call incorrectly returned **0** at O1 and O2. Dead-local elimination replaced the entire `ShortCircuit` with `Goto(join)` when its result was unused, deleting the conditional mutation. The emitted function was just:

```text
load_var box
load_field .value
return
```

The fix replaces the discarded logical result with an ordinary conditional branch, retaining RHS execution. After the PR:

```text
load_var a
pop_jump_if_false L0
load_var2 2 2
load_field .value
load_const 1
bin_op +
store_field .value
L0:
load_var box
load_field .value
return
```

Here both operands of `load_var2 2 2` refer to `box`. Regression tests cover AND/OR, field/index writes, and captured locals in `ns_optimizer_effects/optimizer_effects.baml`.

### Unused evaluations also dropped catchable panics

```baml
function unused_division(x: int) -> int {
    { let unused = 1 / x; 42 } catch (e) {
        baml.panics.DivisionByZero => 7
    }
}

test "unused evaluation must still panic" {
    assert.equal(unused_division(0), 7)
}
```

Previously this returned **42** because the division was deleted. The first commit adds conservative discardability checks and retains potentially failing evaluations. The regression now returns **7**; unused out-of-bounds indexing is covered too. This correction is a prerequisite for the broader dead-store optimization in the second commit.

## Seven Branch Instructions

The **seven-instruction ordinary branch family is inspired by Python's instruction set**, particularly its distinction between consuming predicate branches and historical value-preserving short-circuit branches. CPython 3.11 documents [`JUMP_IF_FALSE_OR_POP`](https://docs.python.org/3.11/library/dis.html#opcode-JUMP_IF_FALSE_OR_POP) and [`JUMP_IF_TRUE_OR_POP`](https://docs.python.org/3.11/library/dis.html#opcode-JUMP_IF_TRUE_OR_POP); contemporary Python documents consuming [`POP_JUMP_IF_TRUE`](https://docs.python.org/3.14/library/dis.html#opcode-POP_JUMP_IF_TRUE) branches. BAML adapts those ideas to its own CFG and stack VM; these are not seven opcodes copied verbatim from one Python release.

| BAML instruction | Taken edge | Fallthrough edge |
| --- | --- | --- |
| `Jump` | Jump; stack unchanged | Not applicable |
| `PopJumpIfFalse` | Pop false | Pop condition |
| `PopJumpIfTrue` | Pop true | Pop condition |
| `JumpIfFalse` | Keep false | Keep condition |
| `JumpIfFalseOrPop` | Keep false | Pop condition |
| `JumpIfTrueOrPop` | Keep true | Pop condition |
| `JumpIfNotNullOrPop` | Keep non-null value | Pop null |

MIR stays value/CFG-oriented; emission selects the fused, edge-dependent stack behavior after stack-carry validation. This count excludes switch/jump-table, exception, and async control flow. Four variants are appended without renumbering existing discriminants. Artifact ABI version is bumped **2 -> 3** and cache format **11 -> 12**.

## Before/After Bytecode

O2 examples below use the human-readable snapshot notation. Labels, temporary names, and helper namespaces are normalized where noted. Counts are static instructions, not runtime throughput claims.

<details>
<summary>Chained AND: 8 -> 6 instructions, with the first false edge going directly to the final join</summary>

Source: `function chained_and(a: bool, b: bool, c: bool) -> bool { a && b && c }`

Before (`ns_fixtures/ns_short_circuit_locals/bytecode.snap`):

```text
load_var a
jump_if_false L0
pop 1
load_var b
L0:
jump_if_false L1
pop 1
load_var c
L1:
return
```

After:

```text
load_var a
jump_if_false_or_pop L0
load_var b
jump_if_false_or_pop L0
load_var c
L0:
return
```

OR uses `jump_if_true_or_pop` and drops from 10 to 6 instructions. Predicate-only chains instead use popping branches and do not materialize an intermediate logical result.

</details>

<details>
<summary>Mutable destination: 14 -> 10 instructions without weakening phi coverage</summary>

```baml
function conditional_short_circuit(a: bool, b: bool, cond: bool) -> bool {
    let x = false
    if (cond) { x = a && b }
    x
}
```

Before (`ns_fixtures/ns_short_circuit_locals/bytecode.snap`):

```text
load_const false
store_var x
load_var cond
pop_jump_if_false L2
load_var a
jump_if_false L0
pop 1
jump L1
L0:
store_var x
jump L2
L1:
load_var b
store_var x
L2:
load_var x
return
```

After:

```text
load_const false
store_var x
load_var cond
pop_jump_if_false L1
load_var a
jump_if_false_or_pop L0
load_var b
L0:
store_var x
L1:
load_var x
return
```

The expression result can be stack-carried even though `x` must retain a real slot for the outer false edge. MIR now separates those two roles.

</details>

<details>
<summary>Null coalescing: 9 -> 4 instructions and no result slot</summary>

Source: `function nca_plain(n: int?, fallback: int) -> int { n ?? fallback }`

Before (`ns_null_coalesce_assign/bytecode.snap`):

```text
load_var n
store_var_load_var _3
load_const null
cmp_op ==
pop_jump_if_false L0
load_var fallback
store_var _3
L0:
load_var _3
return
```

After:

```text
load_var n
jump_if_not_null_or_pop L0
load_var fallback
L0:
return
```

The existing `x ?? y ?? z` snapshot similarly drops from 16 to 6 instructions.

</details>

<details>
<summary>Immediate virtual-call result: 7 -> 5 instructions</summary>

Source: `function ord_self_in_default(a: Tier, b: Tier) -> bool { a.precedes(b) }`

Before (`ns_operators/bytecode.snap`):

```text
load_var2 1 2
load_type operators.Ordered
load_const "precedes"
virtual_call nargs=2 ntypeargs=0
store_var _0
load_var _0
return
```

After:

```text
load_var2 1 2
load_type operators.Ordered
load_const "precedes"
virtual_call nargs=2 ntypeargs=0
return
```

Loop-header results are not blindly carried: one static use can execute repeatedly. Tests check that a backedge or another incoming edge forces materialization.

</details>

<details>
<summary>Mixed call/constant array prefix: 12 -> 8 instructions and 2 -> 0 temporary slots</summary>

```baml
function number(value: int) -> int { value }
function array_mixed() -> int[] { [number(1), 2, number(3)] }
```

Before (isolated compiler probe; temporary names normalized):

```text
load_const 1
call number
store_var t0
load_const 3
call number
store_var t1
load_var t0
load_const 2
load_var t1
load_type int
alloc_array 3
return
```

After (`ns_optimizer_stack/bytecode.snap`; helper namespace omitted):

```text
load_const 1
call number
load_const 2
load_const 3
call number
load_type int
alloc_array 3
return
```

Only the infallible constant moves before the next call. Effectful evaluations retain their order. Tests cover mutable receivers, noncommutative binary operations, indirect/generic/virtual calls, and unwinding while earlier operands remain on the stack.

</details>

<details>
<summary>Third commit: inlined string concatenation, 8 -> 6 instructions and 1 -> 0 result slots</summary>

```baml
function bt_case_h() -> string {
    let x = "hi"
    `!${x}`
}
```

The second commit moved the literal before `String.from` but retained the result's store/load pair. That particular hunk was reorder-only, not a size improvement:

```text
load_const "!"
load_type string
load_const "hi"
call baml.String.from
store_var _1
load_var _1
bin_op +
return
```

The third commit removes that pair:

```text
load_const "!"
load_type string
load_const "hi"
call baml.String.from
bin_op +
return
```

The existing binary-prefix proof now lives in the shared operand-prefix simulator, so it is used both at direct evaluations and when an expression is pulled through an inlined local into a return or call argument. No MIR or opcode changes, new candidate kinds, or relaxed safety guards are needed: both operands must already be stack-carried in source order, and existing definition, predecessor, depth, and final-plan checks remain intact.

`bt_case_i` similarly drops **10 -> 8** instructions; `bt_case_n` drops **9 -> 7**, both with zero result slots. Exact budgets use the actual backtick fixture source. Additional tests check subtraction/division order, nested expressions, call arguments, side-effect order, and unwinding from both calls and arithmetic at O0/O1/O2. A MIR-to-stack-classification unit test validates both operand depths and rejects reversed stack order.

Comparing this third commit against `045bffe6c` across all **248** snapshots: **22 snapshots shrink by 152 instructions total; no snapshot or individual existing function grows**. `ns_backtick_strings` drops **2,173 -> 2,137** instructions. The new regression tests are Rust integration tests, so they add no BAML corpus registration instructions.

</details>

Across the pre-existing snapshot inventory, displayed instructions decrease from **99,860 to 93,984** (about **5.9%**). This includes test/stdlib scaffolding and is not weighted by execution frequency. No throughput benchmark claim is made.

## Snapshot Regression Audit

Compared all **246 pre-existing bytecode snapshots** against the pre-PR base `048bf98ac`, including every file that grew in the first published revision. Counts below are actual instruction lines, not labels, temporary declarations, or diff churn.

| Snapshot | Pre-PR | Published PR | Revised PR |
| --- | ---: | ---: | ---: |
| `ns_patterns_new_runtime` | 3,772 | 4,329 | 3,570 |
| `ns_array_rest_binding` | 761 | 834 | 743 |
| `ns_classes` | 330 | 338 | 330 |
| `ns_gc` | 76 | 83 | 74 |
| `ns_fixtures/ns_type_annotation` | 23 | 30 | 23 |
| `ns_fixtures/ns_llm_parse_catchable_parse_error` | 110 | 113 | 107 |
| `ns_fixtures/ns_patterns_class_destructure_namespaces` | 110 | 112 | 106 |
| `ns_fixtures/ns_parser_expressions` | 33 | 45 | 38 |
| Root test registry | 649 | 655 | 655 |

The smaller original code in the first seven rows was sound: it omitted unused plain allocations/fields, array projections protected by shape/length guards, or arithmetic bounded by a loop predicate. The correction restores those optimizations through general MIR analysis, with no fixture names, snapshot-specific rules, or hand-restored bytecode:

- Discardability checks include every operand. An unused allocation can disappear, but a potentially failing initializer and initializer calls retain their evaluation.
- Forward CFG facts track integer ranges, container lengths, temporary aliases, and completed reads at each statement. Joins retain only shared facts; reassignment invalidates dependent proofs, heap writes/calls invalidate facts conservatively, and exception handlers start without assumptions from the try body.
- Loop widening bounds analysis time independently of the loop trip count. In-bounds negative and length-relative indices are supported.
- Dead-store/local elimination uses these evaluation-site proofs, not a whole-local purity classification. Temporary discards can be reconsidered when their producer is proven safe.
- Reassigned/defaulted parameter copies can propagate within a block until a source/destination write. Other temporary chains remain available to stackification.

For example, `nested_construction_dead_store_fn` in `ns_classes` regressed from 2 instructions to this 6-instruction sequence in the published PR:

```text
load_const 42
init_instance user.classes.CInner .value
init_instance user.classes.COuter .inner
pop 1
load_const 42
return
```

The revised compiler restores the original 2 instructions:

```text
load_const 42
return
```

The only remaining source-code snapshot increase is **`parser_expressions`: +5 instructions**, a correctness fix. Its old `FieldAccess` omitted `user.tags[0]`; its old `IndexAccess` omitted `users[1]` and `users[0].tags[0]`. Those accesses can panic even when their results are unused. Tests compile the actual fixture sources and check that an empty tags array, a one-element users array, and an empty nested tags array still raise catchable `IndexOutOfBounds` at O0/O1/O2. Non-failing inputs retain the original result.

The root registry's **+6** is test scaffolding: two new regression-test namespaces each add a load/call/pop registration sequence. It is not growth in an unchanged source function.

### Additional CI Coverage

The broader CI suite also exposed a copy-propagation bug: a single-use constant definition could be erased even when its consumer required a `Place`, such as `Len(local)`, where a constant substitution cannot be written. The collector now covers direct place reads as well as projection bases/indices. Existing runtime-session method tests cover the failure, and all 43 runtime-session tests pass.

MIR and inline virtual-call snapshots are refreshed, the artifact-compatibility test follows `FORMAT_VERSION`, and the VM fault-injection test replaces stale PC-indexed metadata along with its instruction stream.

Review also caught an O0 debugging regression: definition-level dead-store elimination now retains overwritten user-named stores at O0 while still optimizing unnamed temporaries. A MIR unit test checks both the individual pass and the full O0 cleanup pipeline, and verifies O1/O2 retain dead-store elimination. The reported virtual-field issue was withdrawn after checking the VM contract: missing interface field links are internal invariant violations, not BAML panics; potentially trapping receiver evaluation remains protected.

## Validation

- Full local nextest suite: **2,073 passed across 88 test binaries, 22 skipped**, on the final revision with snapshot updates disabled.
- [GitHub CI](https://github.com/BoundaryML/baml/actions/runs/33963565684) passed on final head `f793a6484`, including snapshot tests, native/WASM cargo tests, SDK tests, pre-commit, and the cross-platform size gate.
- [CodeRabbit approved](https://github.com/BoundaryML/baml/pull/4759#pullrequestreview-5121076647) final head `f793a6484`.
- Full native BAML corpus: **3,589 passed, 2 intentional tolerated failures, 3,591 total**; aggregate passed. Loopback/network tests were run outside the filesystem/network sandbox.
- `corpus_snapshots` regenerated and then passed with updates disabled.
- New optimizer behavior tests pass at **O0, O1, and O2**, with exact instruction/local-count assertions at O2.
- **58 emitter**, **10 MIR**, **58 VM-types**, **34 VM**, **5 artifact**, and **12 cache** unit tests passed.
- Deterministic emission (5), full/unit linking oracle (5), incremental relinking oracle (8), existing short-circuit regressions (4), loop-reassignment regressions (2), and optimization tests (9) passed.
- Scoped Clippy for the changed compiler/VM/artifact/cache crates, all targets, with `-D warnings`; Rust formatting and `git diff --check` passed.
- The second and third commits' normal pre-commit hooks, including `cargo fmt` and `cargo clippy`, passed.

Local tests used installed Rust **1.97.1** with locked/offline dependencies; initial formatting used installed rustfmt **1.93.0**, and the revised commits' normal formatting hooks used installed rustfmt **1.97.1**. The repository-pinned 1.98.0 toolchain was unavailable locally; GitHub CI supplied the broader platform, SDK, and WASM coverage linked above.

## Summary by CodeRabbit

- **New Features**
  - Added support for null-coalescing (`??`) expressions alongside `&&` and `||`.
  - Added value-preserving short-circuit evaluation and optimized conditional branching.

- **Performance**
  - Improved optimization through constant folding, dead-code elimination, and streamlined bytecode.
  - Reduced unnecessary temporary operations in generated bytecode.

- **Bug Fixes**
  - Preserved behavior for discarded expressions, panic paths, mutable values, virtual calls, and evaluation order.
  - Updated compiled artifact and cache formats to support new runtime instructions.

### CODE REVIEW

**Comment by antoniosarosi** [comment]
@coderabbitai review

Updated head: 85b68be88c58dec26cc2b6aa420cc19461ae1ff1. Please review the new evaluation-site discardability analysis, range/bounds proofs and invalidation at writes/calls/exception entries, and the place-only-read copy-propagation correction. The PR description includes the complete snapshot-growth audit and the remaining panic repros. The two-commit structure is preserved.

**Comment by antoniosarosi** [comment]
@coderabbitai full review

File: baml_language/crates/baml_compiler2_mir/src/ir.rs [hunk 1]
```diff
@@ -1007,6 +1013,81 @@ pub enum Rvalue<'db> {
...
+            | Self::Aggregate {
+                fields: elements, ..
+            } => elements.iter().all(operand),
+            Self::Map(_, _, entries) => entries
+                .iter()
+                .all(|(key, value)| operand(key) && operand(value)),
+            Self::MakeClosure { captures, .. } => captures.iter().all(operand),
+            Self::MakeBoundMethod { receiver, .. } | Self::VirtualFieldAccess { receiver, .. } => {
+                operand(receiver)
+            }
```
[the remark this answers is not in the document]
Comment (antoniosarosi) [reply baml_language/crates/baml_compiler2_mir/src/ir.rs]: @coderabbitai This failure is explicitly an internal invariant violation, not an observable BAML failure: `baml_language/crates/bex_vm/src/vm.rs:4955-4958` documents that the type checker proves the receiver implements the interface and E0124 makes `field_links` total. `resolve_virtual_field_slot` returns `VmInternalError::UnresolvedVirtualFieldAccess` only for a compiler/VM inconsistency. Valid typed MIR therefore permits eliminating an unused virtual field projection, just like a fixed field projection. The receiver still goes through `operand(receiver)`, preserving any potentially trapping indexed receiver evaluation. Moving this into the non-discardable arm would regress sound dead-read elimination without preserving a user-reachable failure. I will document this invariant at the discardability arm; please re-evaluate the finding against the VM contract. I am separately fixing the valid O0 named-store concern.

**Comment by antoniosarosi** [comment]
Review follow-up pushed as 40af69f5, still preserving exactly two commits. The O0 dead-store pass now retains overwritten named-local definitions while optimizing unnamed temporaries; a focused MIR unit test covers O0/O1/O2 and the full O0 cleanup pipeline. The virtual-field finding was withdrawn after checking the existing VM invariant, which is now documented at the discardability arm. The billion-iteration convergence probe is explicitly marked compile-only. The synchronous instruction-budget test remains in baml_tests because it intentionally verifies the complete source-to-linked-bytecode pipeline against real corpus fixtures, not an isolated emitter unit. All six MIR unit tests and normal pre-commit hooks pass; the O2 snapshot audit is unchanged at 94,136 instructions.

**Comment by antoniosarosi** [comment]
@coderabbitai review

File: baml_language/crates/baml_tests/tests/optimizer_effects.rs [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+                .bytecode
+                .instructions
+                .iter()
+                .any(|instruction| matches!(
+                    instruction,
+                    bex_vm_types::Instruction::LoadArrayElement | bex_vm_types::Instruction::AddInt
+                )),
+            "{name}: {:?}",
+            function.bytecode.instructions
+        );
```
[the remark this answers is not in the document]
Comment (antoniosarosi) [reply baml_language/crates/baml_tests/tests/optimizer_effects.rs:271]: @coderabbitai Fixed in 045bffe6: the matcher now rejects `Instruction::BinOp(bex_vm_types::BinOp::Add)` as well as `AddInt` and `LoadArrayElement`. The existing diagnostic output is unchanged. Also addressed both unit-test nits: direct tests cover checked integer arithmetic and negation, zero division/modulo, finite/non-finite float folding, the 4096-byte string limit, and propagation with no intervening write versus source/destination reassignment. All 10 MIR unit tests, all 3 optimizer-effects integration tests, and all 2,071 baml_tests nextest tests pass with snapshot updates disabled. Normal pre-commit hooks pass. This follow-up changes tests only; the production compiler and all snapshots are identical to 40af69f5, whose CI run has completed successfully.

**Comment by antoniosarosi** [comment]
@coderabbitai review

**Comment by antoniosarosi** [comment]
Pushed the separately requested third commit, f793a6484, for binary prefixes inside inlined returns and call arguments. It moves the existing operand-depth proof into the shared simulator without relaxing definition, predecessor, or final-plan checks. The actual backtick fixture now emits 6/8/7 instructions for bt_case_h/i/n with zero result slots. Audit against 045bffe6c: 152 instructions removed across 22 snapshots, with no snapshot or individual function growing. All 58 emitter unit tests and all 2,073 baml_tests nextest tests pass with updates disabled; normal pre-commit hooks pass. The preceding revision completed CI successfully. The PR description now documents all three commits and the new before/after example. Please review the new commit and submit approval if no blocking findings remain.

**Comment by antoniosarosi** [comment]
@coderabbitai review
